src/openrct2-ui/windows/RideConstruction.cpp

```diff
@@ -233,2815 +180,2789 @@ static void CloseRideWindowForConstruction(RideId rideId)
         window_close(w);
 }
 
-/**
- *
- *  rct2: 0x006CB481
- */
-rct_window* WindowRideConstructionOpen()
-{
-    RideId rideIndex = _currentRideIndex;
-    CloseRideWindowForConstruction(rideIndex);
-
-    auto ride = get_ride(rideIndex);
-    if (ride == nullptr)
-        return nullptr;
-
-    _stationConstructed = ride->num_stations != 0;
-    _deferClose = false;
-
-    if (ride->type == RIDE_TYPE_MAZE)
-    {
-        return context_open_window_view(WV_MAZE_CONSTRUCTION);
-    }
-
-    auto w = WindowCreate(
-        ScreenCoordsXY(0, 29), 166, 394, &window_ride_construction_events, WC_RIDE_CONSTRUCTION, WF_NO_AUTO_CLOSE);
-
-    w->widgets = window_ride_construction_widgets;
+static void RideConstructPlacedForwardGameActionCallback(const GameAction* ga, const GameActions::Result* result);
+static void RideConstructPlacedBackwardGameActionCallback(const GameAction* ga, const GameActions::Result* result);
+static void CloseConstructWindowOnCompletion(Ride* ride);
 
-    WindowInitScrollWidgets(w);
+class RideConstructionWindow final : public Window
+{
+private:
+    Ride* _currentRide;
```

Comment:
Unfortunately saving the pointer is dangerous you need to always get the ride from the window number as the pointer may go stale between ticks.

Reply:
@duncanspumpkin am I able to do something like
```
    void OnOpen() override
    {
        rideId = _currentRideIndex;
        auto currentRide = get_ride(rideId);
        ...
    }

    void OnUpdate() override
    {
        auto currentRide = get_ride(rideId);
    ...
```
Or should I always be using `_currentRideIndex`?

Reply:
I'm not entirely sure. I don't think current ride id would change whilst the window is open but i can't be sure. The safest would be to do it exactly as it was before.

Reply:
Sounds good, this PR is ready for another look then.